### PR TITLE
Upgrade to v8

### PR DESCRIPTION
**Pull request by kentdr**
Update to the latest version of [Amazon.SQS transport](https://github.com/Particular/NServiceBus.amazonsqs)

### CODE REVIEW

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs [hunk 1]
```diff
@@ -21,27 +21,47 @@ public class AwsLambdaSQSEndpointConfiguration
...
+                .CustomPolicy(recoverabilityPolicy.Invoke);
+
+            if (sqsClient is null && snsClient is null)
+            {
+                // If both sqsClient/snsClient are null, use default constructor which will instantiate a default client instance.
+                Transport = new SqsTransport();
+            }
+            else
+            {
+                // If either sqsClient or snsClient is null the transport will throw.
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs:52]: isn't that what the test setup (at the very beginning of this PR) does?

File: src/NServiceBus.AwsLambda.SQS/LambdaLog.cs [hunk 2]
```diff
@@ -4,6 +4,8 @@ namespace NServiceBus.AwsLambda.SQS
     using Amazon.Lambda.Core;
     using NServiceBus.Logging;
 
+    using LogLevel = Logging.LogLevel;
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/LambdaLog.cs:7]: why was this necessary?

File: src/NServiceBus.AwsLambda.SQS/NServiceBus.AwsLambda.SQS.csproj [hunk 3]
```diff
@@ -1,15 +1,15 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFrameworks>net6.0</TargetFrameworks>
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/NServiceBus.AwsLambda.SQS.csproj:4]: we should probably switch to `TargetFramework` if we only target 1

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs [hunk 4]
```diff
@@ -1,26 +1,53 @@
...
 
-        public override TransportInfrastructure Initialize(SettingsHolder settings, string connectionString)
+        public override async Task<TransportInfrastructure> Initialize(HostSettings hostSettings, ReceiveSettings[] receivers, string[] sendingAddresses, CancellationToken cancellationToken = default)
         {
-            var baseTransportInfrastructure = baseTransport.Initialize(settings, connectionString);
-            return new ServerlessTransportInfrastructure<TBaseTransport>(baseTransportInfrastructure, settings);
+            //   hostSettings.
+            var baseTransportInfrastructure = await BaseTransport.Initialize(hostSettings, receivers, sendingAddresses, cancellationToken).ConfigureAwait(false);
+            var serverlessTransportInfrastructure = new ServerlessTransportInfrastructure<TBaseTransport>(baseTransportInfrastructure);
+            PipelineInvoker = (PipelineInvoker)serverlessTransportInfrastructure.Receivers[MainReceiverId];
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:30]: this will cause issues if we're trying to enable some `SendOnly` like scenarios where there is no main receiver

File: src/NServiceBus.AwsLambda.SQS/LambdaLog.cs (hunk 2 above)
Comment (kentdr) [reply src/NServiceBus.AwsLambda.SQS/LambdaLog.cs:7]: There is a naming conflict between `Amazon.Lambda.Core.LogLevel` and `NServiceBus.Logging.LogLevel`

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs (hunk 4 above)
Comment (kentdr) [reply src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:30]: Is there a test case I can set up?

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs (hunk 1 above)
Comment (kentdr) [reply src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs:52]: I fixed the tests to pass an SNS Client.  I thought the transport threw if the client is null, but after further review,  the SqsTransport does not throw.  Though in my opinion, the behavior of the transport on this is a bit unpredictable.  If you call the default constructor it will initialize them to default implementations, if you call this other one, it does not check for null nor does it initialize default implementations.  I'm not sure if sending null in for these has a predictable outcome and fear it's a bit of a time bomb for when you find out if sending in null is an ok scenario.

File: src/NServiceBus.AwsLambda.SQS/LambdaLog.cs (hunk 2 above)
Comment (timbussmann) [reply src/NServiceBus.AwsLambda.SQS/LambdaLog.cs:7]: was that changed with v8?

File: src/NServiceBus.AwsLambda.SQS/TransportMessage.cs [hunk 5]
```diff
@@ -14,7 +13,7 @@ public TransportMessage()
         {
         }
 
-        public TransportMessage(OutgoingMessage outgoingMessage, List<DeliveryConstraint> deliveryConstraints)
+        public TransportMessage(OutgoingMessage outgoingMessage, List<DispatchProperties> deliveryConstraints)
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/TransportMessage.cs:16]: doesn't need to be a list I think (see https://github.com/Particular/NServiceBus.AmazonSQS/blob/master/src/NServiceBus.Transport.SQS/TransportMessage.cs)

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs (hunk 4 above)
Comment (timbussmann) [reply src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:30]: I've tried to create some here https://github.com/Particular/NServiceBus.AzureFunctions.InProcess.ServiceBus/pull/618 but I think we can scope this out of this PR for now. Would `SendOnly` work on the previous version or would that throw an exception too?

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransportInfrastructure.cs [hunk 6]
```diff
@@ -1,56 +1,27 @@
...
             this.baseTransportInfrastructure = baseTransportInfrastructure;
-            this.settings = settings;
+            Dispatcher = baseTransportInfrastructure.Dispatcher;
+            Receivers = baseTransportInfrastructure.Receivers.ToDictionary(r => r.Key, r => (IMessageReceiver)new PipelineInvoker(r.Value));
         }
 
-        public override IEnumerable<Type> DeliveryConstraints =>
-            baseTransportInfrastructure.DeliveryConstraints;
-
+        //TODO: Is this still necessary?
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransportInfrastructure.cs:17]: no, this can be removed

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransportInfrastructure.cs [hunk 7]
```diff
@@ -1,56 +1,27 @@
...
-    using System.Collections.Generic;
-    using Routing;
-    using Settings;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
     using Transport;
 
-    class ServerlessTransportInfrastructure<TBaseInfrastructure> : TransportInfrastructure
+    class ServerlessTransportInfrastructure<TBaseInfrastructure> : TransportInfrastructure where TBaseInfrastructure : TransportDefinition
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransportInfrastructure.cs:8]: is the generic needed? `TBaseInfrastructure` shouldn't be of type `TransportDefinition` but I think this doesn't seem to be needed at all.

File: src/NServiceBus.AwsLambda.SQS.AcceptanceTests/NServiceBus.AwsLambda.SQS.AcceptanceTests.csproj [hunk 8]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp3.1;net7.0</TargetFrameworks>
+    <TargetFramework>net6.0</TargetFramework>
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS.AcceptanceTests/NServiceBus.AwsLambda.SQS.AcceptanceTests.csproj:4]:
```suggestion
    <TargetFrameworks>net6.0;net7.0</TargetFrameworks>
```

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs [hunk 9]
```diff
@@ -52,30 +72,30 @@ void TrySpecifyDefaultLicense()
...
+
+        /// <summary>
+        /// Amazon SQS Transport
+        /// </summary>
+        public SqsTransport Transport { get; }
 
         internal EndpointConfiguration EndpointConfiguration { get; }
-        internal PipelineInvoker PipelineInvoker { get; private set; }
+
+        internal ServerlessTransport<SqsTransport> MakeServerless()
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs:86]: this is currently not being called by the testing setups from what I can tell, so they might not be correctly configured.

**Review by timbussmann — CHANGES_REQUESTED** [review]

File: src/NServiceBus.AwsLambda.SQS.AcceptanceTests/AwsLambdaSQSEndpointTestBase.cs [hunk 10]
```diff
@@ -133,6 +136,12 @@ public static IAmazonSQS CreateSQSClient()
             return new AmazonSQSClient(credentials);
         }
 
+        public static IAmazonSimpleNotificationService CreateSNSClient()
```
Comment (danielmarbach) [inline src/NServiceBus.AwsLambda.SQS.AcceptanceTests/AwsLambdaSQSEndpointTestBase.cs:139]: If we bump the minimum required dependency of the SDK types to versions that reference [3.7.106 of AWSSDK.Core](https://www.nuget.org/packages/AWSSDK.Core/3.7.106) or higher we can leave the settings as is. Otherwise we would have to introduce a client factory similar to

https://github.com/Particular/NServiceBus.AmazonSQS/blob/master/src/NServiceBus.Transport.SQS.Tests/ClientFactories.cs

File: src/NServiceBus.AwsLambda.SQS.AcceptanceTests/NServiceBus.AwsLambda.SQS.AcceptanceTests.csproj (hunk 8 above)
Comment (danielmarbach) [reply src/NServiceBus.AwsLambda.SQS.AcceptanceTests/NServiceBus.AwsLambda.SQS.AcceptanceTests.csproj:4]: Agreed. Also check the github action and drop 3.1 https://github.com/Particular/NServiceBus.AwsLambda.Sqs/blob/master/.github/workflows/ci.yml#L42

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs [hunk 11]
```diff
@@ -295,20 +297,18 @@ async Task ProcessMessageWithInMemoryRetries(Dictionary<string, string> headers,
...
+                        new Dictionary<string, string>(headers),
+                        body,
+                        transportTransaction,
+                        queueUrl,
+                        new ContextBag());
 
-                        await Process(messageContext, lambdaContext).ConfigureAwait(false);
+                    await Process(messageContext, lambdaContext, token).ConfigureAwait(false);
+
+                    messageProcessedOk = !token.IsCancellationRequested;
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs:310]: I think we should as much as possible align this code with the InputQueuePump of the first major version of the SQS transport that was released for v8

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs [hunk 12]
```diff
@@ -172,21 +174,21 @@ public async Task Unsubscribe(Type eventType, ILambdaContext lambdaContext)
...
+
+            s3BucketForLargeMessages = configuration.Transport.S3?.BucketName;
             if (string.IsNullOrWhiteSpace(s3BucketForLargeMessages))
             {
                 return;
             }
-
-            var s3ClientFactory = settingsHolder.GetOrDefault<Func<IAmazonS3>>(SettingsKeys.S3ClientFactory) ?? (() => new AmazonS3Client());
-            s3Client = s3ClientFactory();
+            s3Client = configuration.Transport.S3?.S3Client;
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs:191]: Maybe because of https://github.com/Particular/NServiceBus.AwsLambda.Sqs/pull/246/files#diff-beaa942efe8e6241de1fcd1b3d60cb8c77dbae5096e343635fc5b4e8ff169781R47 ?

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs [hunk 13]
```diff
@@ -21,27 +21,47 @@ public class AwsLambdaSQSEndpointConfiguration
...
+            {
+                // If either sqsClient or snsClient is null the transport will throw.
+                Transport = new SqsTransport(sqsClient, snsClient);
+            }
+
+            RoutingSettings = EndpointConfiguration.UseTransport(Transport);
 
             // by default do not write custom diagnostics to file because lambda is readonly
-            AdvancedConfiguration.CustomDiagnosticsWriter(diagnostics => Task.CompletedTask);
+            AdvancedConfiguration.CustomDiagnosticsWriter((diagnostics, token) => Task.CompletedTask);
```
Comment (danielmarbach) [inline src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs:59]: We could use the discard `_` operator here for both values

File: src/NServiceBus.AwsLambda.SQS/NServiceBus.AwsLambda.SQS.csproj [hunk 14]
```diff
@@ -1,15 +1,15 @@
...
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>net6.0</TargetFramework>
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>..\NServiceBus.snk</AssemblyOriginatorKeyFile>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="Amazon.Lambda.Core" Version="[2.1.0, 3.0.0)" />
     <PackageReference Include="Amazon.Lambda.SQSEvents" Version="[2.1.0, 3.0.0)" />
```
Comment (danielmarbach) [inline src/NServiceBus.AwsLambda.SQS/NServiceBus.AwsLambda.SQS.csproj:11]: Can we make sure to bump the lowest required to the latest?

File: src/NServiceBus.AwsLambda.SQS/TransportMessage.cs (hunk 5 above)
Comment (danielmarbach) [reply src/NServiceBus.AwsLambda.SQS/TransportMessage.cs:16]: The whole empty message body and body preparation process probably needs an overhaul. We have done quite a bit of changes to make native integration possible. That being said, we can likely do this as subsequent PRs

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/PipelineInvoker.cs [hunk 15]
```diff
@@ -1,30 +1,40 @@
...
+        public string ReceiveAddress => baseTransportReceiver.ReceiveAddress;
+
+        Task IMessageReceiver.Initialize(PushRuntimeSettings limitations, OnMessage onMessage, OnError onError, CancellationToken cancellationToken)
+        {
+            this.onMessage = onMessage;
+            this.onError = onError;
+
+            return baseTransportReceiver?.Initialize(limitations,
+                (_, __) => Task.CompletedTask,
+                (_, __) => Task.FromResult(ErrorHandleResult.Handled),
```
Comment (danielmarbach) [inline src/NServiceBus.AwsLambda.SQS/TransportWrapper/PipelineInvoker.cs:28]: This task could be cached

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransportInfrastructure.cs (hunk 7 above)
Comment (danielmarbach) [reply src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransportInfrastructure.cs:8]: Agreed the generic is not required

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs [hunk 16]
```diff
@@ -1,26 +1,53 @@
 ﻿namespace NServiceBus.AwsLambda.SQS.TransportWrapper
 {
-    using Settings;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NServiceBus;
     using Transport;
 
     class ServerlessTransport<TBaseTransport> : TransportDefinition
```
Comment (danielmarbach) [inline src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:9]: Could we also remove the generic here?

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs [hunk 17]
```diff
@@ -1,26 +1,53 @@
...
+#pragma warning disable CS0672 // Member overrides obsolete member
+#pragma warning disable CS0618 // Type or memeber is obsolete
+
+        public override string ToTransportAddress(QueueAddress address) => BaseTransport.ToTransportAddress(address);
+
+#pragma warning restore CS0618 // Type or memeber is obsolete
+#pragma warning restore CS0672 // Member overrides obsolete member
+
+
+        public override IReadOnlyCollection<TransportTransactionMode> GetSupportedTransactionModes() => supportedTransactionModes;
```
Comment (danielmarbach) [inline src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:45]: Would it be possible to delegate this forward to the base transport or is this dangerous because we might need to write explicit code in the lambda support to support a tx mode? I think it is quite unlikely though that we get SendsWithAtomicReceive with SQS / SNS

Comment (timbussmann) [reply src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:45]: not sure forwarding makes sense, even when the values are the same anyway. The transaction support is heavily determined by the pump which we don't control here, so the transaction modes listed here are what the trigger/hosting model allow us to use/integrate with, not what our SQS transport implementation could do.

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs (hunk 4 above)
Comment (kentdr) [reply src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:30]: I spiked a bit on this and got something functional.  I put up a PR once this one is merged.

File: src/NServiceBus.AwsLambda.SQS.AcceptanceTests/AwsLambdaSQSEndpointTestBase.cs (hunk 10 above)
Comment (kentdr) [reply src/NServiceBus.AwsLambda.SQS.AcceptanceTests/AwsLambdaSQSEndpointTestBase.cs:139]: Is your suggestion based on having the ability to set the HttpClientCacheSize?

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs (hunk 9 above)
Comment (kentdr) [reply src/NServiceBus.AwsLambda.SQS/AwsLambdaSQSEndpointConfiguration.cs:86]: [AwsLambdaEndpoint calls it automatically during initialization ](https://github.com/Particular/NServiceBus.AwsLambda.Sqs/blob/upgrade-to-v8/src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs#L63)

I did debug the tests and confirmed that it does get called.

File: src/NServiceBus.AwsLambda.SQS.AcceptanceTests/AwsLambdaSQSEndpointTestBase.cs (hunk 10 above)
Comment (danielmarbach) [reply src/NServiceBus.AwsLambda.SQS.AcceptanceTests/AwsLambdaSQSEndpointTestBase.cs:139]: The cache would be correct as of that version and we can remove the explicit setting to one. Same applies to the transport and dynamodb by the way.

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs (hunk 17 above)
Comment (danielmarbach) [reply src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:44]: Agreed. My. Thought was coming from the angle that the transport pump and this pump should always be as close as possible

Comment (kentdr) [reply src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:44]: Do we agree then to leave this as is?

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs [hunk 18]
```diff
@@ -285,30 +282,29 @@ static bool IsMessageExpired(SQSEvent.SQSMessage receivedMessage, Dictionary<str
...
+                        body,
+                        transportTransaction,
+                        queueUrl,
+                        new ContextBag());
 
-                        messageProcessedOk = !messageContextCancellationTokenSource.IsCancellationRequested;
-                    }
+                    await Process(messageContext, lambdaContext, token).ConfigureAwait(false);
+
+                    messageProcessedOk = true;
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs:307]: can't we just return here instead?

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs [hunk 19]
```diff
@@ -324,7 +320,10 @@ async Task ProcessMessageWithInMemoryRetries(Dictionary<string, string> headers,
                             nativeMessageId,
                             body,
                             transportTransaction,
-                            immediateProcessingAttempts);
+                            immediateProcessingAttempts,
+                            queueUrl,
+                            new ContextBag()
+                            );
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs:326]:
```suggestion
                            new ContextBag());
```

File: src/NServiceBus.AwsLambda.SQS/TransportMessage.cs [hunk 20]
```diff
@@ -14,7 +11,7 @@ public TransportMessage()
         {
         }
 
-        public TransportMessage(OutgoingMessage outgoingMessage, List<DeliveryConstraint> deliveryConstraints)
+        public TransportMessage(OutgoingMessage outgoingMessage, DispatchProperties deliveryConstraints)
```
Comment (timbussmann) [inline src/NServiceBus.AwsLambda.SQS/TransportMessage.cs:14]: can we rename the argument name to `properties` so that it is aligned with the transport's type?

File: src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs (hunk 17 above)
Comment (danielmarbach) [reply src/NServiceBus.AwsLambda.SQS/TransportWrapper/ServerlessTransport.cs:44]: Good for me

File: src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs [hunk 21]
```diff
@@ -229,7 +226,7 @@ async Task ProcessMessage(SQSEvent.SQSMessage receivedMessage, ILambdaContext la
             }
             catch (OperationCanceledException)
             {
-                return;
```
[the remark this answers is not in the document]
Comment (kentdr) [reply src/NServiceBus.AwsLambda.SQS/AwsLambdaEndpoint.cs:232]: Lambda isn't passing a cancellation token into the function, the caller of Process is supplying it or passing default.  In this case, the only thing that would throw an `OperationCancelledException` is the call to `transportMessage.RetrieveBody`  which is before any handlers are called, so, throwing out, in my opinion, is the right call.  If we just return here, then the message will be dequeued unless the caller of Process returns an error code or throws, which is a loss scenario unless I'm mistaken.    These are the types of error cases I want to test some more but that will be on a follow-up PR.

The Lambda context has a "RemainingTime" property that is passed in.  I don't know what happens to the message if that time expires before the function returns, something else to test.

File: src/NServiceBus.AwsLambda.SQS.AcceptanceTests/AwsLambdaSQSEndpointTestBase.cs (hunk 10 above)
Comment (kentdr) [reply src/NServiceBus.AwsLambda.SQS.AcceptanceTests/AwsLambdaSQSEndpointTestBase.cs:157]: I bumped to the latest SQS transport, is that sufficient?  We don't currently have a direct dependency on AWSSDK.Core .

**Review by danielmarbach — APPROVED** [review]
